app/jobs/sync_pardot_prospects_and_subscriptions_job.rb

```diff
@@ -0,0 +1,46 @@
+class SyncPardotProspectsAndSubscriptionsJob < ApplicationJob
+  queue_as :default
+  sidekiq_options retry: 2
+
+  def perform(salesforce_ids = [])
+    # without an argument, this processes all contacts - else you can specify an array of salesforce_ids to process
+    query = Contact.all
+    query = query.where(salesforce_id: salesforce_ids) unless salesforce_ids.empty?
+
+    query.each do |contact|
+      current_subscriptions = Subscription.includes(:list).where(contact: contact)
+      new_subscriptions = []
+
+      prospect = Pardot.client.prospects.read_by_fid(contact.salesforce_id)
+      next unless prospect['lists']
+
+      prospect['lists']['list_subscription'].each do |subscription|
+        if subscription.include?('list')
+          # Returns an array if it only has one subscription, but a hash otherwise. O_o This seems to handle it nicely.
+          subscription = if subscription.is_a?(Array)
+                           subscription[1]
+                         else
+                           subscription['list']
+                         end
+
+          next unless subscription['is_public'] == 'true'
+
+          new_subscriptions.append(subscription['id'].to_i)
+          list = List.where(pardot_id: subscription['id']).first_or_create do |plist|
+            plist.title = subscription['title']
+            plist.description = subscription['description']
+          end
```

Comment:
this code is making lists that don't exist - am I reading it correctly that this will be called many times for the same list?  Is there another way to do this where we first sync all lists, and then sync subscriptions?  Also, what happens when a list is deleted from pardot - is that handled somewhere in the syncing?

---- 
actually reading more below this code exists in the "update list" code -- so is there some code to DRY up?

Reply:
I merged the lists/subscription job into one `PardotSyncJob`, there was no reason to have these as separate.